### PR TITLE
[WIP] feat: add autofocus logic to the engine

### PR DESCRIPTION
**Pull request by fdrgsp**
This PR adds to the `MDAEngine` the logic to control any Micromanager Autofocus device (e.g. Nikon PFS).

Linked to https://github.com/pymmcore-plus/useq-schema/pull/91.

(test will fail until useq-schema PR is merged)

### CODE REVIEW

File: src/pymmcore_plus/mda/_engine.py [hunk 1]
```diff
@@ -32,6 +58,10 @@ def setup_sequence(self, sequence: MDASequence) -> None:
 
         self._mmc = self._mmc or CMMCorePlus.instance()
 
+        # switch off autofocus device if it is on
+        with contextlib.suppress(RuntimeError):
+            self._mmc.setProperty(self._mmc.getAutoFocusDevice(), "State", "Off")
```
Comment (ianhi) [inline src/pymmcore_plus/mda/_engine.py:63]: why not:

`self._mmc.setContinuousFocus(False)`

File: src/pymmcore_plus/mda/_engine.py [hunk 2]
```diff
@@ -59,6 +115,18 @@ def exec_event(self, event: MDAEvent) -> Any:
...
+        """Perform the autofocus.
+
+        Returns the correction to be applied to the focus motor position.
+        """
+        self._mmc.setPosition(
+            autofocus_event.autofocus_z_device_name,
+            cast(float, autofocus_event.af_motor_offset),
+        )
+        self._mmc.fullFocus()
+        return self._mmc.getZPosition() - cast(float, autofocus_event.z_stage_position)
```
Comment (ianhi) [inline src/pymmcore_plus/mda/_engine.py:128]: I've found that there absolutely need to be some waits in here. either `waitForSystem` or `waitForDevices`.

On some of my older systems i've also found it be significantly more robust to catch an error in autofocus and retry once or twice. Anecdotally on the olympus system I use 1 retry works most of the time, and 2 retries has yet to fail even over a 16 hour timelapse.

Here's the equivalent section of my engine:

https://github.com/ianhi/raman-mda-engine/blob/f7238bc286aa9642331eb5c6215c458e31327115/raman_mda_engine/_engine.py#L273-L284

Comment (ianhi) [reply src/pymmcore_plus/mda/_engine.py:128]: > I've found that there absolutely need to be some waits in here.

without the waits on some systems the `getZPosition` may not report on the correctly focused stage position.

File: src/pymmcore_plus/mda/_engine.py (hunk 1 above)
Comment (fdrgsp) [reply src/pymmcore_plus/mda/_engine.py:63]: @ianhi you meant `self._mmc.enableContinuousFocus(False)`? Sure, it should work as well!

**Comment by tlambert03** [comment]
closing because I think #204 is preferred now?

**Pull request closed without merging**